airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/context/env/ConnectorConfigurationPropertySource.kt

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.core.context.env
+
+import com.fasterxml.jackson.core.type.TypeReference
+import io.airbyte.cdk.integrations.base.JavaBaseConstants
+import io.airbyte.commons.json.Jsons
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.env.MapPropertySource
+import io.micronaut.core.cli.CommandLine
+import io.micronaut.core.util.StringUtils
+import java.nio.file.Path
+import java.util.AbstractMap
+import java.util.stream.Collectors
+import java.util.stream.Stream
+
+private val logger = KotlinLogging.logger {}
+
+/**
+ * Custom Micronaut {@link PropertySource} that reads the command line arguments provided via the
+ * connector CLI and turns them into configuration properties. This allows the arguments to be
+ * injected into code that depends on them via Micronaut. <p /> This property source adds the
+ * following properties to the configuration if the matching value is present in the CLI arguments:
+ * <ul> <li><b>airbyte.connector.operation</b> - the operation argument (e.g. "check", "discover",
+ * etc)</li> <li><b>airbyte.connector.catalog</b> - the Airbyte configured catalog as JSON read from
+ * the configured catalog file path argument, if present.</li> <li><b>airbyte.connector.config</b> -
+ * the normalized Airbyte connector configuration read from the configured connector configuration
+ * file path argument, if present.</li> <li><b>airbyte.connector.state</b> - the Airbyte connector
+ * state as JSON read from the state file path argument, if present.</li> </ol>
+ */
+class ConnectorConfigurationPropertySource(commandLine: CommandLine) :
+    MapPropertySource("connector", resolveValues(commandLine)) {
```

Comment:
Should we implement our own PropertySource? A map is a bit too free-form to my liking.

Reply:
@postamar Property sources in Micronaut are not to populate an object directly, but rather to turn some source into properties in Micronaut (hence the map).  From there, you can map the properties to an object, like I have done in this PR.